hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieFileIndex.scala

```diff
@@ -144,22 +146,41 @@ case class HoodieFileIndex(spark: SparkSession,
     // Prune the partition path by the partition filters
     // NOTE: Non-partitioned tables are assumed to consist from a single partition
     //       encompassing the whole table
-    val prunedPartitions = listMatchingPartitionPaths(partitionFilters)
+    val prunedPartitions = if (shouldBroadcast) {
+      listMatchingPartitionPaths(convertFilterForTimestampKeyGenerator(metaClient, partitionFilters))
+    } else {
+      listMatchingPartitionPaths(partitionFilters)
+    }
     val listedPartitions = getInputFileSlices(prunedPartitions: _*).asScala.toSeq.map {
       case (partition, fileSlices) =>
-        val baseFileStatuses: Seq[FileStatus] = getBaseFileStatus(fileSlices
+        var baseFileStatuses: Seq[FileStatus] = getBaseFileStatus(fileSlices
           .asScala
-          .map(fs => fs.getBaseFile.orElse(null))
-          .filter(_ != null))
-
+            .map(fs => fs.getBaseFile.orElse(null))
+            .filter(_ != null))
+        if (shouldBroadcast) {
+          baseFileStatuses = baseFileStatuses ++ fileSlices.asScala
+            .filter(f => f.getLogFiles.findAny().isPresent && !f.getBaseFile.isPresent)
+            .map(f => f.getLogFiles.findAny().get().getFileStatus)
+        }
         // Filter in candidate files based on the col-stats index lookup
         val candidateFiles = baseFileStatuses.filter(fs =>
           // NOTE: This predicate is true when {@code Option} is empty
           candidateFilesNamesOpt.forall(_.contains(fs.getPath.getName)))
 
         totalFileSize += baseFileStatuses.size
         candidateFileSize += candidateFiles.size
-        PartitionDirectory(InternalRow.fromSeq(partition.values), candidateFiles)
+        if (shouldBroadcast) {
+          val c = fileSlices.asScala.filter(f => f.getLogFiles.findAny().isPresent
+            || (f.getBaseFile.isPresent && f.getBaseFile.get().getBootstrapBaseFile.isPresent)).
+            foldLeft(Map[String, FileSlice]()) { (m, f) => m + (f.getFileId -> f) }
+          if (c.nonEmpty) {
+            PartitionDirectory(new PartitionFileSliceMapping(InternalRow.fromSeq(partition.values), spark.sparkContext.broadcast(c)), candidateFiles)
+          } else {
+            PartitionDirectory(InternalRow.fromSeq(partition.values), candidateFiles)
+          }
+        } else {
+          PartitionDirectory(InternalRow.fromSeq(partition.values), candidateFiles)
+        }
```

Comment:
Could you move the logic to a single if branch when broadcast is enabled, so it's easier to read?

Reply:
I tried that before but don't think it looks better. I did it again and you can take a look